File: Algorithm/python_source/charpter18_backtracking/level1/has_path_sum.py
# coding=utf-8
from python_source.tool.Tree import Tree

#  判断树是否一样
class HasPathSum:
    def pathSum(self, root, targetSum):
        ret = list()
        path = list()

        def dfs(root, targetSum):
            if not root:
                return
            path.append(root.val)
            targetSum -= root.val
            if not root.left and not root.right and targetSum == 0:
                ret.append(path[:])
            dfs(root.left, targetSum)
            dfs(root.right, targetSum)
            path.pop()

        dfs(root, targetSum)
        return ret


if __name__ == "__main__":
    tree = Tree()

    root = tree.init_tree_for_path_sum()
    hasPathSum = HasPathSum()
    res = hasPathSum.hasPathSum(root,22)
    print res

File: Algorithm/Java/src/main/java/com/yugutou/charpter18_backtracking/level1/HasPathSum.java
package com.yugutou.charpter18_backtracking.level1;


import com.yugutou.tools.BinaryTree;
import com.yugutou.tools.TreeNode;

import java.util.Deque;
import java.util.LinkedList;
import java.util.List;

/**
 * LeetCode 112
 * 给你二叉树的根节点 root 和一个表示目标和的整数 targetSum 。
 * 判断该树中是否存在 根节点到叶子节点 的路径，这条路径上所有节点值相加等于目标和 targetSum 。
 * 如果存在，返回 true ；否则，返回 false 。
 */
public class HasPathSum {
    public static void main(String[] args) {
        BinaryTree bTree = new BinaryTree();
        bTree.root = bTree.buildBinaryTree();

        boolean result1 = hasPathSum(bTree.root, 12);
        System.out.println(result1);

        List<List<Integer>> pathSum = pathSum(bTree.root, 12);
        System.out.println(pathSum);

    }


    /**
     * 问题2：LeetCode112. 路径总和
     *
     * @param root
     * @param sum
     * @return
     */
    public static boolean hasPathSum(TreeNode root, int sum) {
        if (root == null) {
            return false;
        }
        if (root.left == null && root.right == null) {
            return sum == root.val;
        }
        return hasPathSum(root.left, sum - root.val) || hasPathSum(root.right, sum - root.val);
    }


    /**
     * 问题3  LeetCode113 路径总和II
     */
    static List<List<Integer>> ret = new LinkedList<List<Integer>>();
    static Deque<Integer> path = new LinkedList<Integer>();

    public static List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        dfs(root, targetSum);
        return ret;
    }

    public static void dfs(TreeNode root, int targetSum) {
        if (root == null) {
            return;
        }
        path.offerLast(root.val);
        targetSum -= root.val;
        if (root.left == null && root.right == null && targetSum == 0) {
            ret.add(new LinkedList<Integer>(path));
        }
        dfs(root.left, targetSum);
        dfs(root.right, targetSum);
        path.pollLast();
    }

}
